src/wallet/wallet_tx_builder.cpp

```diff
@@ -177,14 +206,190 @@ SpendableInputs WalletTxBuilder::FindAllSpendableInputs(
+        }
+        logicalActionCount = CalculateLogicalActionCount(
+                vins,
+                vouts,
+                std::max(inputs.value().sproutNoteEntries.size(), sproutOutputCount),
+                inputs.value().saplingNoteEntries.size(),
+                NotOne(saplingOutputCount),
+                NotOne(std::max(inputs.value().orchardNoteMetadata.size(), orchardOutputCount)));
+
+    } else {
+        logicalActionCount = CalculateLogicalActionCount(
+                {},
+                vouts,
+                sproutOutputCount,
+                0,
+                NotOne(saplingOutputCount),
+                NotOne(orchardOutputCount));
+    }
+
+    return CalculateConventionalFee(logicalActionCount);
+}
+
+InvalidFundsError ReportInvalidFunds(
+        const SpendableInputs& spendable,
+        bool hasQuasiChange,
+        CAmount fee,
+        CAmount dustThreshold,
+        CAmount targetAmount,
+        CAmount changeAmount)
+{
+    return InvalidFundsError(
+            spendable.Total(),
+            hasQuasiChange
+            // TODO: NEED TESTS TO EXERCISE THIS
+            ? InvalidFundsReason(QuasiChangeError(fee, dustThreshold))
+            : (changeAmount > 0 && changeAmount < dustThreshold
+               // TODO: we should provide the option for the caller to explicitly forego change
+               // (definitionally an amount below the dust amount) and send the extra to the
+               // recipient or the miner fee to avoid creating dust change, rather than prohibit
+               // them from sending entirely in this circumstance. (Daira disagrees, as this could
+               // leak information to the recipient or publicly in the fee.)
+               ? InvalidFundsReason(DustThresholdError(dustThreshold, changeAmount))
+               : InvalidFundsReason(InsufficientFundsError(targetAmount))));
+}
+
+/// On the initial call, we haven’t yet selected inputs, so we assume the outputs dominate the
+/// actions.
+///
+/// 1. calc fee using only resolvedPayments to set a lower bound on the actual fee
+///    • this also needs to know which pool change is going to, so it can determine what the fee is
+///      with change _if_ there is change
+/// 2. iterate over LimitToAmount until the updated fee (now including spends) matches the expected
+///    fee
+tl::expected<std::pair<SpendableInputs, CAmount>, InputSelectionError>
+IterateLimit(
+        CAmount sendAmount,
+        CAmount dustThreshold,
+        const SpendableInputs& spendable,
+        const Payments& resolved,
+        const std::optional<std::optional<OutputPool>>& changePool)
+{
+    SpendableInputs spendableMut;
+
+    auto previousFee = MINIMUM_FEE;
+    auto updatedFee = CalcZIP317Fee(std::nullopt, resolved.GetResolvedPayments(), std::nullopt);
+    // This is used to increase the target amount just enough (generally by 0 or 1) to force
+    // selection of additional notes.
+    CAmount bumpTargetAmount{0};
```

Comment:
Rather than solely relying on the target amount, it seems like note selection should be made aware of any "free spends" that we have available as a consequence of the resolved payments. In particular, if a transaction already has a privacy policy that permits outputs across multiple shielded pools, then it would make sense to select inputs across those same shielded pools as well.

Non-blocking; this would just be a fee optimization.